### PR TITLE
Mitigation for inactive pixel regions for phase1 tracking

### PR DESCRIPTION
**Pull request by makortel**
This PR adds pixel pair iteration to phase1 tracking for
- backup for triplets in the barrel-forward transition not covered by 4 layers (as in #17891 for phase 2)
- mitigation for the region currently covered only by two active barrel layers
  * seeding from layers BPix1+BPix4

In order to keep fakes and timing under control, the regions are restricted to first 5 `firstStepPrimaryVertices`. As a knock-on effect, `firstStepPrimaryVertices` are now sorted (similar to `offlnePrimaryVertices`). The sorting will likely cause tiny, numerical-level effect on phase0 tracking.

MTV plots for phase1 ttbar+PU35 are here
https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre3_PR19051/
comparing 910pre3 out of the box (all modules active), recent pixel bad module map, pixelPair iteration w/o the mitigation part, and the full PR.

Tested in 9_1_0_pre3 and CMSSW_9_2_X_2017-05-30-1100. Expecting changes as shown above in phase1, numerical-level changes on phase0 (caused by sorting `firstStepPrimaryVertices`), and no effect on phase2.

@rovere @VinInn @mtosi @JanFSchulte

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: FastSimulation/Configuration/python/Reconstruction_AftMix_cff.py [hunk 1]
```diff
@@ -64,15 +64,16 @@
...
-# so we replace the input collection for firstStepPrimaryVertices with generalTracks
-_reco.firstStepPrimaryVertices.TrackLabel = "generalTracks"
+# so we replace the input collection for firstStepPrimaryVerticesUnsorted with generalTracks
+_reco.firstStepPrimaryVerticesUnsorted.TrackLabel = "generalTracks"
+_reco.initialStepTrackRefsForJets.src = "generalTracks"
 
 # insert the few tracking modules to be run after mixing back in the globalreco sequence
 #for _entry in reversed([trackExtrapolator,caloTowerForTrk,firstStepPrimaryVertices,ak4CaloJetsForTrk])
-_reco.globalreco.insert(0,_reco.trackExtrapolator+_reco.caloTowerForTrk+_reco.firstStepPrimaryVertices+_reco.ak4CaloJetsForTrk)
+_reco.globalreco.insert(0,_reco.trackExtrapolator+_reco.caloTowerForTrk+_reco.firstStepPrimaryVerticesUnsorted+_reco.ak4CaloJetsForTrk+_reco.initialStepTrackRefsForJets+_reco.firstStepPrimaryVertices)
```
Comment (makortel) [inline FastSimulation/Configuration/python/Reconstruction_AftMix_cff.py]: I'm not really sure about how I dealt with FastSim when I sorted the `firstStepPrimaryVertices`. What I did is the `firstStepPrimaryVerticesBeforeMixing` in initialStep are unsorted, while the `firstStepPrimaryVertices` here (after mixing) are sorted. I tried to sort them also in initialStep, but that lead to circular dependencies on the modules (firstStepPrimaryVerticesBeforeMixing requiring ak4CaloJetsForTrkBeforeMixing, that requires caloTowerForTrkBeforeMixing, that requires ecalRecHit, that requires simEcalDigis, that requires simEcalTriggerPrimitiveDigis, that requires mix), so without knowing much of the detailes here I decided to proceed with an easy way.

Could FastSim experts take a look? 

@ssekmen @ANSH0712

**Comment by slava77** [comment]
@makortel 
if this PR was covered in a presentation in TRK POG or in the DPG meeting, please add a link in the description for improved self-documentation
Thank you.

**Comment by dmitrijus** [comment]
+1

**Comment by perrotta** [comment]
A  few plots follows showing the amount of changes in tracking related variables in phase1 (TTBar sample with PU, wf 10224: taking also into account that superimposed PU events were not the same, the effects are those anticipated in the description, and detailed for the tracking quantities in the link provided).  About phase0 and phase2 already the jenkins tests show respectively tiny and no changes, as anticipated in the PR description

Eta of slimmed pat electrons:
image

Eta of slimmed jets:
image

MET sumEt:
image

Muon eta:
image

**Comment by perrotta** [comment]
Timing was  already shown in the link attached to the description (baseline is the first column, the full PR is the last one):

image

**Comment by perrotta** [comment]
+1
The PR does what expected, according to jenkins tests, the tracking specific plots provided in https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre3_PR19051/ , and https://github.com/cms-sw/cmssw/pull/19051#issuecomment-306831970

There is still pending the question raised by Matti in https://github.com/cms-sw/cmssw/pull/19051#pullrequestreview-41353977, but this relates to FastSim and can be dealt with by the dedicated reviewer

**Comment by slava77** [comment]
On 6/7/17 8:31 AM, perrotta wrote:
> into account that superimposed PU events were not the same

why were they not the same?
Please rerun from the same step2 inputs to have a more consistent comparison

**Comment by slava77** [comment]
On 6/7/17 8:35 AM, perrotta wrote:
> Timing was already shown in the link attached

this describes only the tracking sequence.
Does everything downstream look OK?

**Comment by slava77** [comment]
it's important to check that b-tagging is OK.
Inclusive distributions posted earlier without MC truth requirements are only indications that nothing is "completely" broken (10% efficiency changes can easily hide if these reco comparison plots were the only things posted).
Andrea, please post a plot, or perhaps Matti has one already and could share.

**Comment by perrotta** [comment]
hold
(while the new runs requested by Slava goes on during the night...)

**Comment by slava77** [comment]
@makortel 
looking at 
https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre3_PR19051/phase1_ttbar_pu35_ootb/effandfakePosDeltaRPU.pdf
the efficiency recovery is restricted to just the first bin  in r position (r< 1 cm) and the loss remains through the pixel volume.
Is this due to the restriction in z vertex?
Was it concluded to be not worth it from point of view of fake rate?
(if it was just timing, then perhaps this restriction needs to be lifted to recover efficiency).

**Comment by VinInn** [comment]
it looks to me a side effect (higher is the PU more detached track we collect even if we ask vertex pointing)
please check 0PU

**Comment by slava77** [comment]
On 6/7/17 10:07 AM, Vincenzo Innocente wrote:
> it looks to me a side effect (higher is the PU more detached track we
> collect even if we ask vertex pointing)
> please check 0PU

Is the efficiency MTV plot using the TPs from the ttbar or is it the
"AllTPs" kind of plot?
For the former, I'm not sure if reference to 0PU is informative.

**Comment by makortel** [comment]
@slava77 Your link
https://cms-tracking-validation.web.cern.ch/cms-tracking-validation/PR/CMSSW_9_1_0_pre3_PR19051/phase1_ttbar_pu35_ootb/effandfakePosDeltaRPU.pdf
points to an efficiency plot using only ttbar signal TPs (I see I didn't include the "signal + in-time PU TPs" variant to this set of plots).

**Comment by perrotta** [comment]
Retested with the wf 10024 (phase1, no PU this time). Still, the modifications are in according to what anticipated in the description of this PR: A few comparisons take as example follow:


- pt of PF candidates:
all_testpr19051vsorig_ttbar13tev2017wf10024p0c_log10recopfcandidates_particleflow__reco_obj_pt

- ak4 pf jets et:
all_testpr19051vsorig_ttbar13tev2017wf10024p0c_log10recopfjets_ak4pfjets__reco_obj_et

- eta of gsf electrons
all_testpr19051vsorig_ttbar13tev2017wf10024p0c_recogsftracks_electrongsftracks__reco_obj_eta

- MET pt
all_mini_testpr19051vsorig_ttbar13tev2017wf10024p0c_patmets_slimmedmets__reco_obj_pt

- B-tag distributions looks mostly only slightly affected, the largest differences being observed in the following plot (which shows that performance improves with this PR: red is new and black is the baseline):
btag

- There is a reduction of pixel-less tracks in the barrel, concentrated around phi=PI:
pixellesseta
pixellessphi

compensated by the other itarations:
trackingalgos

for an overall improved efficiency:
trackingeff

**Comment by perrotta** [comment]
Timing tested on the same sample show an overall increase of about 30ms/evt:
- Job total:  2.66811 s/ev ==> 2.69607 s/ev

where the new modules contribute for the following amount:

       added      +0.01%         0.00 ms/ev ->         0.21 ms/ev pixelPairStep
       added      +0.00%         0.00 ms/ev ->         0.09 ms/ev initialStepTrackRefsForJets
       added      +0.00%         0.00 ms/ev ->         0.04 ms/ev pixelPairStepSeedLayersB
       added      +0.00%         0.00 ms/ev ->         0.08 ms/ev pixelPairStepClusters
       added      +0.01%         0.00 ms/ev ->         0.18 ms/ev pixelPairStepSeedsA
       added      +0.00%         0.00 ms/ev ->         0.04 ms/ev pixelPairStepSeedsB
       added      +0.00%         0.00 ms/ev ->         0.03 ms/ev cutsRecoTracksPixelPairStep
       added      +0.12%         0.00 ms/ev ->         3.31 ms/ev pixelPairStepTrackCandidates
       added      +0.07%         0.00 ms/ev ->         1.78 ms/ev pixelPairStepTracks
       added      +0.00%         0.00 ms/ev ->         0.03 ms/ev cutsRecoTracksPixelPairStepHp
       added      +0.02%         0.00 ms/ev ->         0.60 ms/ev TrackSeedMonpixelPairStep
       added      +0.00%         0.00 ms/ev ->         0.02 ms/ev pixelPairStepTrackingRegions
       added      +0.00%         0.00 ms/ev ->         0.06 ms/ev pixelPairStepHitDoubletsB
       added      +0.00%         0.00 ms/ev ->         0.07 ms/ev pixelPairStepSeedLayers
       added      +0.01%         0.00 ms/ev ->         0.25 ms/ev pixelPairStepHitDoublets
       added      +0.00%         0.00 ms/ev ->         0.02 ms/ev pixelPairStepSeeds
       added      +0.01%         0.00 ms/ev ->         0.26 ms/ev pixelPairStepTrackingRegionsB
       added      +0.14%         0.00 ms/ev ->         3.62 ms/ev firstStepPrimaryVerticesUnsorted

**Comment by perrotta** [comment]
unhold

**Comment by makortel** [comment]
Minor detail, but I wonder how `initialStepTrackRefsForJets` can be a "new module" given that it is already run as part of `JetCoreRegionalStep`?
https://github.com/cms-sw/cmssw/blob/master/RecoTracker/IterativeTracking/python/JetCoreRegionalStep_cff.py#L188
I only moved it to `InitialStep`. (all other modules are indeed new)

**Comment by perrotta** [comment]
Indeed... It looks like the comparison tool marks as "added" also the modules which takes 0 CPU in the baseline (I think within the resolution) in the baseline:
- baseline        ->  initialStepTrackRefsForJets 0.000000
- testPR19051 -> initialStepTrackRefsForJets 9.595e-05

**Comment by slava77** [comment]
On 6/8/17 3:22 AM, perrotta wrote:
> B-tag distributions looks mostly only slightly affected, the largest
> differences being observed in the following plot (which shows that
> performance improves with this PR: red is new and black is the baseline):

very nice.
May I ask to please add either a comb MVAv2 or CSVv2 ROC plot.
Thank you.

**Comment by slava77** [comment]
On 6/8/17 3:27 AM, perrotta wrote:
>   * Job total: 2.66811 s/ev ==> 2.69607 s/ev

is this without PU? (should be with PU)

Is this total without VALIDATION sequence ? (it should be without)

**Comment by slava77** [comment]
On 6/8/17 3:22 AM, perrotta wrote:
> Retested with the wf 10024 (phase1, no PU this time).

no-PU choice here simplifies the tracking task.
So, problems that may be present with PU will likely look smaller.

**Comment by perrotta** [comment]
image

**Comment by perrotta** [comment]
csvv2

**Comment by perrotta** [comment]
In the timing comparison I did NOT remove the validation sequence

**Comment by slava77** [comment]
On 6/8/17 5:45 AM, perrotta wrote:
> csvv2
> <https://user-images.githubusercontent.com/4069749/26929207-181acf9c-4c59-11e7-9381-9f6a65ff0fda.jpg>
> 

this is c.vs.b
does the light.vs.b look good (more important, IMO)

**Comment by perrotta** [comment]
csvv2_vslightq

**Pull request merged**